content/blog/2024/09/usr-merge/index.md

```diff
@@ -0,0 +1,90 @@
+---
+title: "Usr Merge"
+author: "evan"
+categories:
+  - news
+date: 2024-09-04
+featuredimage: "/usr-merge-banner-image.jpg"
+url: "/2024/09/04/usr-merge"
+---
+
+Back in May, 2024, we made our initial attempt at doing what we're calling "Usr-Merge". While initially hopeful, it quickly became clear that there were significant problems with our approach, and that a new one would be needed. This post will go into what happened in May, and outline what we plan to do next time.
+
+## What is Usr-Merge?
+
+The old way of organizing the Linux filesystem split files between `/bin` and `/usr/bin`, `/sbin` and `/usr/sbin`, `/lib` and `/usr/lib`, and `/lib64` and `/usr/lib64`. In recent years, there has been a push to unify everything by putting all the directories in the `/usr` file tree, e.g. `/usr/bin` and `/usr/lib64`, and adding compatibility symbolic links (symlinks) pointing to the directories in `/usr`. There are several reasons for this:
+
+- It improves compatibility in behavior with other Unixes and Linuxes. Scripts written for another distribution don't need modifications for different file paths, making them more portable.
+- Compatibility with the GNU build systems. Porting software to Linux becomes easier, because the build system and developers don't have to worry about a split `/usr`.
+- Improved compatibility with upstream development.
+- Reduced complexity of the filesystem. The separation between vendor-provided (coming from your Linux distribution, possibly read-only) resources and user resources becomes more clear.
+
+The Freedesktop Wiki has a much more detailed article about Usr-Merge [here](https://www.freedesktop.org/wiki/Software/systemd/TheCaseForTheUsrMerge/).
+
+### Sounds like Stateless
+
+If you've used Solus for any length of time, you may have heard the team talking about "stateless" packages. What does that mean? A stateless system means that, in theory, you can remove everything in `/etc` and be left with a still-working system using vendor-provided (e.g., Solus) defaults. We do this by putting default configuration files in `/usr/share/defaults`, which can be copied to `/etc` or `$HOME/.config` and modified by the user, overwriting the system defaults. While not directly related to Statelessness, the idea behind it and Usr-Merge are quite similar.
+
+This concept is also called "Hermetic-Usr" by systemd and openSUSE.
+
+Serpent OS and `moss` takes the idea of Stateless a step further. The package manager *cannot* install files outside of `/usr`, enforcing a stateless system. Since Solus will also be moving to `moss` and associated tooling, we will also not be able to ship files outside of `/usr`, including files in directories like `/bin` and `/lib`. That's where Usr-Merge comes in; it is a prerequisite to switching to the Serpent tooling on Solus.
```

Comment:
```suggestion
Serpent OS and `moss` take the idea of Statelessness a step further. The package manager *cannot* install files outside of `/usr`, enforcing a stateless system. Since Solus will also be moving to `moss` and associated tooling, we will also not be able to ship files outside of `/usr`, including files in directories like `/bin` and `/lib`. That's where Usr-Merge comes in; it is a prerequisite to switching to the Serpent OS tooling on Solus.
```